File: Backend/Nodejs/src/models/playlist.ts
import { pool } from "./database/connection";
import { Playlist, UpdatePlaylist } from "./types";

export class PlaylistModel {
  static createPlaylist(
    playlist: Playlist,
    cover: Express.MulterS3.File,
    callback: (response: any, ok: Boolean) => void,
  ) {
    try {
      const newPlaylist = { ...playlist };
      newPlaylist.cover = cover.location;
      console.log(newPlaylist);
      pool.query("CALL CreatePlaylist(?,?,?,?)", [
        newPlaylist.name,
        newPlaylist.description,
        newPlaylist.cover,
        newPlaylist.email,
      ], (err, result) => {
        if (err) throw err;
        if (result[0][0].TYPE == "ERROR") callback(result[0][0], false);
        else {
          callback(result[0][0], true);
        }
      });
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }

  static getUserPlaylist(
    { email }: { email: string },
    callback: (response: any, ok: Boolean) => void,
  ) {
    try {
      pool.query("CALL GetUserPlaylists(?)", [
        email,
      ], (err, result) => {
        if (err) throw err;
        callback(result[0], true);
      });
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }

  static getPlaylist(
    { id }: { id: string },
    callback: (response: any, ok: Boolean) => void,
  ) {
    try {
      pool.query(
        `SELECT id_playlist AS id, name, description, image AS cover, email
        FROM PR1.Playlists
        WHERE id_playlist=?`,
        id,
        (err, result) => {
          if (err) throw err;
          callback(result[0], true);
        },
      );
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }

  static getSongs(
    { id }: { id: string },
    callback: (response: any, ok: Boolean) => void,
  ) {
    try {
      pool.query("CALL GetPlaylistSongs(?)", [
        id,
      ], (err, result) => {
        if (err) throw err;
        callback(result[0], true);
      });
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }


  static getSongsNotInPlaylist(
    { id }: { id: string },
    callback: (response: any, ok: Boolean) => void,
  ) {
    try {
      pool.query("CALL GetSongsNotInPlaylist(?)", [
        id,
      ], (err, result) => {
        if (err) throw err;
        callback(result[0], true);
      });
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }




  static addSong(
    playlist: number,
    song: number,
    email: string,
    callback: (response: any, ok: Boolean) => void,
  ) {
    try {
      pool.query("CALL AddSongPlaylist(?,?,?)", [
        playlist,
        song,
        email,
      ], (err, result) => {
        if (err) throw err;
        if (result[0][0].TYPE == "ERROR") callback(result[0][0], false);
        else {
          callback(result[0][0], true);
        }
      });
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }

  static removeSong(
    playlist: number,
    song: number,
    email: string,
    callback: (response: any, ok: Boolean) => void,
  ) {
    try {
      pool.query("CALL RemoveSongPlaylist(?,?,?)", [
        playlist,
        song,
        email,
      ], (err, result) => {
        if (err) throw err;
        if (result[0][0].TYPE == "ERROR") callback(result[0][0], false);
        else {
          callback(result[0][0], true);
        }
      });
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }

  static editPlaylist(
    newPlaylist: UpdatePlaylist,
    file: Express.MulterS3.File,
    callback: Function,
  ) {
    try {
      newPlaylist.cover = file ? file.location : "";
      console.log(newPlaylist)
      // Guardar en DB
      pool.query("CALL UpdatePlaylist(?,?,?,?,?)", [
        newPlaylist.id,
        newPlaylist.name,
        newPlaylist.description,
        newPlaylist.cover,
        newPlaylist.email,
      ], (err, result) => {
        if (err) throw err;
        if (result[0][0].TYPE == "ERROR") callback(result[0][0], false);
        else {
          callback(result[0][0], true);
        }
      });
    } catch (error) {
    }
  }

  static deletePlaylist(
    { id, email }: { id: number; email: string },
    callback: (response: any, ok: Boolean) => void,
  ) {
    try {
      pool.query("CALL RemovePlaylist(?,?)", [
        id,
        email,
      ], (err, result) => {
        if (err) throw err;
        if (result[0][0].TYPE == "ERROR") callback(result[0][0], false);
        else {
          callback(result[0][0], true);
        }
      });
    } catch (error) {
      console.log(error);
      callback(error, false);
    }
  }
}

File: Backend/Python/src/models/playlist.py
from src.models.database.connection import getCnx


class PlaylistModel:

    @staticmethod
    def create_playlist(playlist, cover_filename):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "CreatePlaylist",
                (
                    playlist["name"],
                    playlist["description"],
                    cover_filename,
                    playlist["email"],
                ),
            )
            db.commit()
            for result in cursor.stored_results():
                result = dict(zip(result.column_names, result.fetchone()))
                if result["TYPE"] == "ERROR":
                    cursor.close()
                    return result, False
                else:
                    cursor.close()
                    return result, True
        except Exception as e:
            cursor.close()
            return str(e), False

    @staticmethod
    def get_users_playlists(email):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "GetUserPlaylists",
                (email,),
            )
            data = []
            for result in cursor.stored_results():
                for playlist in result.fetchall():
                    # Falta el cover
                    result = dict(zip(("id", "cover", "name", "description"), playlist))
                    data.append(result)
            cursor.close()
            return data, True
        except Exception as e:
            cursor.close()
            return str(e), False

    @staticmethod
    def get_songs(id):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "GetPlaylistSongs",
                (id,),
            )
            data = []
            for result in cursor.stored_results():
                for song in result.fetchall():
                    # Falta el cover
                    result = dict(zip(("id", "name", "cover", "musicSrc"), song))
                    data.append(result)
            cursor.close()
            return data, True
        except Exception as e:
            cursor.close()
            return str(e), False

    @staticmethod
    def songs_not_playlist(id):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "GetSongsNotInPlaylist",
                (id,),
            )
            data = []
            for result in cursor.stored_results():
                for song in result.fetchall():
                    # Falta el cover
                    result = dict(
                        zip(("id", "name", "cover", "musicSrc", "singer"), song)
                    )
                    data.append(result)
            cursor.close()
            return data, True
        except Exception as e:
            cursor.close()
            return str(e), False

    @staticmethod
    def add_song(playlist, song, email):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "AddSongPlaylist",
                (playlist, song, email),
            )
            db.commit()
            for result in cursor.stored_results():
                result = dict(zip(result.column_names, result.fetchone()))
                if result["TYPE"] == "ERROR":
                    cursor.close()
                    return result, False
                else:
                    cursor.close()
                    return result, True
        except Exception as e:
            cursor.close()
            return str(e), False

    @staticmethod
    def remove_song(playlist, song, email):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "RemoveSongPlaylist",
                (playlist, song, email),
            )
            db.commit()
            for result in cursor.stored_results():
                result = dict(zip(result.column_names, result.fetchone()))
                if result["TYPE"] == "ERROR":
                    cursor.close()
                    return result, False
                else:
                    cursor.close()
                    return result, True
        except Exception as e:
            cursor.close()
            return str(e), False

    @staticmethod
    def remove_playlist(id, email):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Query
            cursor.callproc(
                "RemovePlaylist",
                (id, email),
            )
            db.commit()
            for result in cursor.stored_results():
                result = dict(zip(result.column_names, result.fetchone()))
                if result["TYPE"] == "ERROR":
                    cursor.close()
                    return result, False
                else:
                    cursor.close()
                    return result, True
        except Exception as e:
            cursor.close()
            return str(e), False

    @staticmethod
    def edit_playlist(playlist, cover_filename):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            print(playlist, cover_filename)
            # Query
            cursor.callproc(
                "UpdatePlaylist",
                (
                    playlist["id"],
                    playlist["name"],
                    playlist["description"],
                    cover_filename,
                    playlist["email"],
                ),
            )
            db.commit()
            for result in cursor.stored_results():
                result = dict(zip(result.column_names, result.fetchone()))
                if result["TYPE"] == "ERROR":
                    cursor.close()
                    return result, False
                else:
                    cursor.close()
                    return result, True
        except Exception as e:
            print(e)
            cursor.close()
            return str(e), False

    @staticmethod
    def get_playlist(id):
        try:
            db = getCnx()  # Obtiene una conexión desde la función
            cursor = db.cursor(buffered=True)
            # Ejecuta la consulta SQL para obtener el usuario por su email
            cursor.execute(
                f"""SELECT id_playlist AS id, name, description, image AS cover, email
        FROM PR1.Playlists
        WHERE id_playlist={id}"""
            )
            result = cursor.fetchall()
            # Convierte el resultado en un diccionario y lo devuelve
            response = []
            for obj in result:
                album = dict(zip(cursor.column_names, obj))
                response.append(album)
            cursor.close()
            return response, True
        except Exception as e:
            cursor.close()
            return str(e), False
